omegaconf/omegaconf.py

```diff
@@ -101,43 +100,62 @@ def SI(interpolation: str) -> Any:
 
 
 def register_default_resolvers() -> None:
-    def env(key: str, default: Any = _EMPTY_MARKER_) -> Any:
+    # DEPRECATED: remove in 2.2
+    def legacy_env(key: str, default: Optional[str] = None) -> Any:
+        warnings.warn(
+            "The `env` resolver is deprecated, see https://github.com/omry/omegaconf/issues/573",
+        )
+
         try:
-            val_str = os.environ[key]
+            return decode_primitive(os.environ[key])
         except KeyError:
-            if default is not _EMPTY_MARKER_:
-                return default
+            if default is not None:
+                return decode_primitive(default)
             else:
                 raise ValidationError(f"Environment variable '{key}' not found")
 
-        # We obtained a string from the environment variable: we try to parse it
-        # using the grammar (as if it was a resolver argument), so that expressions
-        # like numbers, booleans, lists and dictionaries can be properly evaluated.
-        try:
-            parse_tree = parse(
-                val_str, parser_rule="singleElement", lexer_mode="VALUE_MODE"
+    def env(key: str, default: Any = _EMPTY_MARKER_) -> Optional[str]:
```

Comment:
```suggestion
    def env(key: str, default: Optional[str] = _EMPTY_MARKER_) -> Optional[str]:
```

Reply:
Done in dfcbf44
(had to make `_EMPTY_MARKER_` a string -- there's a neligible risk of someone accidentally using that same string which I think should be fine -- if you don't like it I can either create an instance of a custom subclass of `str` or make `default` a `Union`)